File: src/test/java/ownDataRepresentation/fileSystem/LatLonDirTest.java
package ownDataRepresentation.fileSystem;

import de.esymetric.jerusalem.ownDataRepresentation.fileSystem.LatLonDir;
import org.junit.Test;

import static org.junit.Assert.assertTrue;

public class LatLonDirTest {

    @Test
    public void testLanLonDir() throws Exception {
        for (double lat = -89; lat < 89; lat += 0.5)
            for (double lng = -89; lng < 89; lng += 0.3) {
                boolean ok = new LatLonDir(lat, lng).equals(new LatLonDir(
                        new LatLonDir(lat, lng).getShortKey()));
                assertTrue(ok);
                // System.out.println("" + lat + " " + lng + " " + ok);
            }

        checkOffsetBits(48, 11, 48, 11);

        checkOffsetBits(48, 11, 49, 11);
        checkOffsetBits(48, 11, 49, 12);
        checkOffsetBits(48, 11, 48, 12);
        checkOffsetBits(48, 11, 48, 10);

        checkOffsetBits(48, 11, 47, 10);
        checkOffsetBits(48, 11, 47, 12);
        checkOffsetBits(48, 11, 47, 11);
        checkOffsetBits(48, 11, 49, 10);

    }


    private void checkOffsetBits(int lat1, int lng1, int lat2, int lng2) {
        LatLonDir lld1 = new LatLonDir(lat1, lng1);
        int offsetBits = lld1.getOffsetBits(lat2, lng2);
        LatLonDir lld3 = new LatLonDir(lat1, lng1, offsetBits);
        assertTrue(new LatLonDir(lat2, lng2).equals(lld3));
    }

}

File: src/test/java/ownDataRepresentation/fileSystem/LatLonDirTest.kt
package ownDataRepresentation.fileSystem

import de.esymetric.jerusalem.ownDataRepresentation.fileSystem.LatLonDir
import junit.framework.Assert.assertNull
import org.junit.Assert
import org.junit.Test

class LatLonDirTest {
    @Test
    @Throws(Exception::class)
    fun testLanLonDir() {
        var lat = -89.0
        while (lat < 89) {
            var lng = -89.0
            while (lng < 89) {
                val ok = LatLonDir(lat, lng).equals(
                    LatLonDir(
                        LatLonDir(lat, lng).shortKey
                    )
                )
                Assert.assertTrue(ok)
                lng += 0.3
            }
            lat += 0.5
        }
        checkOffsetBits(48, 11, 48, 11)
        checkOffsetBits(48, 11, 49, 11)
        checkOffsetBits(48, 11, 49, 12)
        checkOffsetBits(48, 11, 48, 12)
        checkOffsetBits(48, 11, 48, 10)
        checkOffsetBits(48, 11, 47, 10)
        checkOffsetBits(48, 11, 47, 12)
        checkOffsetBits(48, 11, 47, 11)
        checkOffsetBits(48, 11, 49, 10)
        checkOffsetBitsNotNeighbouring(48, 11, 46, 11)
        checkOffsetBitsNotNeighbouring(48, 11, 49, 13)
    }

    private fun checkOffsetBits(lat1: Int, lng1: Int, lat2: Int, lng2: Int) {
        val lld1 = LatLonDir(lat1.toDouble(), lng1.toDouble())
        val offsetBits = lld1.getOffsetBits(lat2.toDouble(), lng2.toDouble())
        val lld3 = LatLonDir(lat1.toDouble(), lng1.toDouble(), offsetBits!!)
        Assert.assertTrue(LatLonDir(lat2.toDouble(), lng2.toDouble()).equals(lld3))
    }

    private fun checkOffsetBitsNotNeighbouring(lat1: Int, lng1: Int, lat2: Int, lng2: Int) {
        val lld1 = LatLonDir(lat1.toDouble(), lng1.toDouble())
        val offsetBits = lld1.getOffsetBits(lat2.toDouble(), lng2.toDouble())
        assertNull(offsetBits)
    }

}